runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,502 @@
+        or args.data.method == ms.textDocument_didOpen
+      )
+    then
+      refresh(bufnr, { client_id = client.id })
+    end
+  end,
+})
+
+---@param kind lsp.DocumentHighlightKind
+---@return string
+local function hl_group_from_kind(kind)
+  if kind == 2 then
+    return 'LspReferenceRead'
+  elseif kind == 3 then
+    return 'LspReferenceWrite'
+  else -- kind == 1 also the default
+    return 'LspReferenceText'
+  end
+end
+
+api.nvim_set_decoration_provider(namespace, {
+  on_win = function(_, _, bufnr, toprow, botrow)
+    local bufstate = bufstates[bufnr]
+    if not bufstate then
+      return
+    end
+
+    if bufstate.version == bufstate.applied then
+      return
+    end
+
+    local enabled = bufstate.enabled
+    if enabled == nil then
+      enabled = globalstate.enabled
+    end
+
+    if not enabled then
+      return
+    end
+
+    local client_highlights = bufstate.client_highlights
+
+    for row = toprow, botrow do
+      api.nvim_buf_clear_namespace(bufnr, namespace, row, row + 1)
+
+      for _, row_highlights in pairs(client_highlights) do
+        local highlights = row_highlights[row] or {}
+
+        for _, highlight in pairs(highlights) do
+          local col = highlight.range['start'].character
+          local end_row = highlight.range['end'].line
+          local end_col = highlight.range['end'].character
+
+          api.nvim_buf_set_extmark(bufnr, namespace, row, col, {
+            end_row = end_row,
+            end_col = end_col,
+            ephemeral = false,
+            hl_group = hl_group_from_kind(highlight.kind),
+          })
+        end
+      end
+    end
+
+    bufstate.applied = bufstate.version
+  end,
+})
+
+---Optional filters |kwargs|, or `nil` for all.
+---@class vim.lsp.document_highlight.enable.Filter
+---@inlinedoc
+---
+---Buffer number, or 0 for current buffer, or nil for all.
+---@field bufnr? integer
+
+---Query whether document highlight is enabled in the {filter}ed scope
+---@param filter? vim.lsp.document_highlight.enable.Filter
+---@return boolean
+function M.is_enabled(filter)
+  vim.validate({ filter = { filter, 'table', true } })
+  filter = filter or {}
+
+  local bufnr = filter.bufnr
+  if bufnr == nil then
+    return globalstate.enabled
+  end
+
+  bufnr = bufnr == 0 and api.nvim_get_current_buf() or bufnr
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return false
+  end
+
+  if bufstate.enabled == nil then
+    return globalstate.enabled
+  else
+    return bufstate.enabled
+  end
+end
+
+---Enables or disables document highlights for the {filter}ed scope.
+---
+---To "toggle", pass the inverse of `is_enabled()`:
+---
+---```lua
+---vim.lsp.document_highlight.enable(not vim.lsp.document_highlight.is_enabled())
+---```
+---@param enable? boolean
+---@param filter? vim.lsp.document_highlight.enable.Filter
+function M.enable(enable, filter)
+  vim.validate('enable', enable, 'boolean', true)
+  vim.validate('filter', filter, 'table', true)
```

Comment:
just call vim.validate once, `enable` and `filter` can be in a table

Reply:
No, this would revert a recent PR from @lewis6991. This was deliberate. See #30855.

Reply:
Yes, it was suggested by lewis [there](https://github.com/neovim/neovim/pull/30841#discussion_r1806116877)

Reply:
np i just see the code without see more context.